contracts/pcv/maker/MakerRouter.sol

```diff
@@ -0,0 +1,170 @@
+        daiPSM.buyGem(to, (minDaiAmountOut) / 1e12);
+    }
+
+    /// @notice Function to swap all of FEI balance to USDC
+    /// @param to the address the USDC should be sent to once swapped
+    function swapAllFeiForUsdc(address to)
+        external
+        hasAnyOfTwoRoles(TribeRoles.GOVERNOR, TribeRoles.PCV_CONTROLLER)
+    {
+        uint256 minDaiAmountOut = _redeemAllBalanceFromFeiPSM(address(this));
+        daiPSM.buyGem(to, (minDaiAmountOut) / 1e12);
+    }
+
+    /// @notice Function to swap for both DAI and USDC
+    /// @dev Function will swap from FEI to DAI first then DAI to USDC
+    /// @param amountFeiIn the amount of FEI to be deposited
+    /// @param minDaiAmountOut the minimum amount of DAI expected to be received
+    /// @param ratioUSDC the ratio of the DAI received we would like to swap to USDC - in basis point terms
+    /// @param to the address the DAI should be sent to once swapped
+    function swapFeiForUsdcAndDai(
+        uint256 amountFeiIn,
+        uint256 minDaiAmountOut,
+        uint256 ratioUSDC,
+        address to
+    )
+        external
+        hasAnyOfTwoRoles(TribeRoles.GOVERNOR, TribeRoles.PCV_CONTROLLER)
+    {
+        require(ratioUSDC < 10000, "MakerRouter: USDC Ratio too high");
+
+        _redeemFromFeiPSM(amountFeiIn, minDaiAmountOut, address(this));
+        uint256 usdcAmount = (minDaiAmountOut * ratioUSDC) / 10000;
+
+        daiPSM.buyGem(to, usdcAmount / 1e12);
+        dai.safeTransfer(to, minDaiAmountOut - usdcAmount);
+    }
+
+    /// @notice Function to swap all FEI balance for both DAI and USDC
+    /// @param to the address the USDC should be sent to once swapped
+    /// @param ratioUSDC the ratio of the DAI received we would like to swap to USDC - in basis point terms
+    function swapAllFeiForUsdcAndDai(address to, uint256 ratioUSDC)
+        external
+        hasAnyOfTwoRoles(TribeRoles.GOVERNOR, TribeRoles.PCV_CONTROLLER)
+    {
+        uint256 minDaiAmountOut = _redeemAllBalanceFromFeiPSM(address(this));
+        uint256 usdcAmount = (minDaiAmountOut * ratioUSDC) / 10000;
+
+        daiPSM.buyGem(to, usdcAmount / 1e12);
+        dai.safeTransfer(to, minDaiAmountOut - usdcAmount);
+    }
+
+    /// @notice Helper function to redeem DAI from the FEI FEI-DAI PSM
+    /// @param amountFeiIn the amount of FEI to be deposited
+    /// @param minDaiAmountOut the minimum amount of DAI expected to be received
+    /// @param to the address the DAI should be sent to once swapped
+    function _redeemFromFeiPSM(
+        uint256 amountFeiIn,
+        uint256 minDaiAmountOut,
+        address to
+    ) private {
+        fei.safeTransferFrom(msg.sender, address(this), amountFeiIn);
+        feiPSM.redeem(to, amountFeiIn, minDaiAmountOut);
+    }
+
+    function _redeemAllBalanceFromFeiPSM(address to)
+        private
+        returns (uint256 minDaiAmountOut)
+    {
+        uint256 amountFeiIn = fei.balanceOf(msg.sender);
```

Comment:
this is subject to a griefing attack if someone sends any more fei to the sender than the sender approves this contract to spend. let's change this line to check the approval of this contract to spend, and then use the min of approval or balance.